### PR TITLE
CreateTeamEngine

### PR DESCRIPTION
**Pull request by oconnor663**
r? @mlsteele @maxtaco 

Trying to follow the example here: https://gist.github.com/maxtaco/2cb8ddbd023a355b85fe0b5babdba927

Things intentionally left out of the code so far:
- Adding multiple users. Right now it just adds the one creator.
- Formatting the `Prev` box. It's always null for the root link, so there's no implementation here.
- Opening the `per_team_key.boxes` format. Nothing does this yet, so their formatting might be wrong, but I eyeballed the output at it at least looks right.

### CODE REVIEW

File: go/libkb/naclwrap.go [hunk 1]
```diff
@@ -79,10 +79,10 @@ type NaclDHKeyPair struct {
...
 }
 
-func (n NaclDHKeyPair) Clone() (ret NaclDHKeyPair) {
-	ret.Public = n.Public
-	if n.Private != nil {
-		tmp := *n.Private
+func (k NaclDHKeyPair) Clone() (ret NaclDHKeyPair) {
+	ret.Public = k.Public
+	if k.Private != nil {
+		tmp := *k.Private
```
Comment (oconnor663) [inline go/libkb/naclwrap.go]: Some linter was complaining about lettering inconsistency here.

File: go/engine/team_create.go [hunk 2]
```diff
@@ -0,0 +1,332 @@
...
+		recipientPerUserNaclKeypair, ok := recipientPerUserGenericKeypair.(libkb.NaclDHKeyPair)
+		if !ok {
+			return nil, fmt.Errorf("got an unexpected key type for recipient KID in sharedTeamKeyBox: %T", recipientPerUserGenericKeypair)
+		}
+		var nonce [24]byte
+		counterBytes := [4]byte{}
+		binary.BigEndian.PutUint32(counterBytes[:], counter)
+		copy(nonce[:20], noncePrefix)
+		copy(nonce[20:24], counterBytes[:])
+		// TODO: pack [ 1, <per-user-seqno>, <nonce-bottom-4-bytes>, <box of current shared team secret> ]
```
Comment (maxtaco) [inline go/engine/team_create.go]: When do we resolve this todo? When we do rotation?

File: go/libkb/kbsig.go [hunk 3]
```diff
@@ -583,7 +583,24 @@ func (u *User) UpdateEmailProof(key GenericKey, newEmail string) (*jsonw.Wrapper
 	return ret, nil
 }
 
-func (u *User) TeamRootSig(key GenericKey, teamSection *jsonw.Wrapper) (*jsonw.Wrapper, error) {
+type TeamSection struct {
+	Name    string `json:"name"`
+	ID      string `json:"id"`
+	Members struct {
+		Owner  []string `json:"owner"`
```
Comment (maxtaco) [inline go/libkb/kbsig.go]: something better than strings here?

File: go/libkb/kbsig.go [hunk 4]
```diff
@@ -583,7 +583,24 @@ func (u *User) UpdateEmailProof(key GenericKey, newEmail string) (*jsonw.Wrapper
 	return ret, nil
 }
 
-func (u *User) TeamRootSig(key GenericKey, teamSection *jsonw.Wrapper) (*jsonw.Wrapper, error) {
+type TeamSection struct {
+	Name    string `json:"name"`
+	ID      string `json:"id"`
```
Comment (maxtaco) [inline go/libkb/kbsig.go]: `keybase1.TeamID`

File: go/libkb/kbsig.go [hunk 5]
```diff
@@ -583,7 +583,24 @@ func (u *User) UpdateEmailProof(key GenericKey, newEmail string) (*jsonw.Wrapper
...
+	Members struct {
+		Owner  []string `json:"owner"`
+		Admin  []string `json:"admin"`
+		Writer []string `json:"writer"`
+		Reader []string `json:"reader"`
+	} `json:"members"`
+	PerTeamKey struct {
+		Generation    int     `json:"generation"`
+		EncryptionKID string  `json:"encryption_kid"`
```
Comment (maxtaco) [inline go/libkb/kbsig.go]: `keybase1.KID`

File: go/libkb/kbsig.go [hunk 6]
```diff
@@ -611,6 +638,10 @@ type SigMultiItem struct {
 	Type       string `json:"type"`
 	SigInner   string `json:"sig_inner"`
 	TeamID     string `json:"team_id"`
+	PublicKeys struct {
```
Comment (maxtaco) [inline go/libkb/kbsig.go]: can we fix this thing to use `keybase1.*` types too?

File: go/engine/team_create.go [hunk 7]
```diff
@@ -0,0 +1,332 @@
...
+	}
+	signingKey, err = libkb.MakeNaclSigningKeyPairFromSecretBytes(derivedSecret(sharedSecret, libkb.TeamEdDSADerivationString))
+	if err != nil {
+		return
+	}
+	return
+}
+
+type PerTeamSharedSecretBoxes struct {
+	Generation    int               `json:"generation"`
```
Comment (maxtaco) [inline go/engine/team_create.go]: introduce a new type for this generation number.

File: go/engine/team_create.go [hunk 8]
```diff
@@ -0,0 +1,332 @@
...
+	EncryptingKid string            `json:"encrypting_kid"`
```
Comment (maxtaco) [inline go/engine/team_create.go]: also use `Keybase1.KID`

File: go/engine/team_create.go (hunk 2 above)
Comment (oconnor663) [inline go/engine/team_create.go]: Woops, I implemented it in the following 3 lines and then forgot to delete the comment.

File: go/engine/team_create.go [hunk 9]
```diff
@@ -0,0 +1,334 @@
...
+		JSONPayload: payload,
+	})
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (e *TeamCreateEngine) getDeviceKeys(ctx *Context, me *libkb.User) (sigKey libkb.NaclSigningKeyPair, encKey libkb.NaclDHKeyPair, err error) {
```
Comment (mlsteele) [inline go/engine/team_create.go]: at some point we gotta factor this out

File: go/engine/team_create.go [hunk 10]
```diff
@@ -0,0 +1,334 @@
...
+	sigSKA := libkb.SecretKeyArg{
```
Comment (patrickxb) [inline go/engine/team_create.go]: You don't need this function at all.

e.G().ActiveDevice.SigningKey()
e.G().ActiveDevice.EncryptionKey()

Comment (oconnor663) [inline go/engine/team_create.go]: ah cool, thanks. will those prompt the user if necessary?

Comment (patrickxb) [inline go/engine/team_create.go]: No because it will never be necessary.  You can't be logged in and not have cached device keys in ActiveDevice.

**Comment by oconnor663** [comment]
Test failure looks like a git clone issue in kbfs-server? https://ci.keyba.se/job/kbfs-server/job/master/12152/execution/node/70/log/

**Pull request merged**